### ISSUE TITLE
Gluon as scripting language

### ISSUE THREAD

**Issue by ShalokShalom** [xi-editor/xi-editor#1189]
Gluon is a statically type inferred scripting language which is embeddable in Rust: 

https://github.com/gluon-lang/gluon

This is potential a superb language to write some expansions, imho

**Comment by cmyr** [comment]
This could definitely be something to explore, I really haven't looked into this much.

My current hunch is that the future of plugins is going to be WASM/WASI, and we've been sort of resistant to bring in a scripting language, but if someone wants to explore that they're very welcome. :)

**Comment by ShalokShalom** [comment]
Yeah, I think its awesome that you can call Rust functions directly. :D

**Comment by erlend-sh** [comment]
> My current hunch is that the future of plugins is going to be WASM/WASI

Amethyst and other Rust gamedev projects are also looking into this: 

https://github.com/amethyst/amethyst/issues/1729
https://github.com/caelunshun/feather/issues/106#issuecomment-529080119

**Comment by MarkMcCaskey** [comment]
@cmyr I'd be happy to help out someone interested in doing this!  At Wasmer we're currently actively working on improving our support for the embedded use case.  As part of that we're working with projects to see where the pain points are so that we can fix them.

Things are moving fast in this area, especially if you want to use WASI, but we do have an in-development API for the embedded WASI use case with features like creating virtual files that implement arbitrary logic via some traits and swapping out files.  In [this example](https://github.com/wasmerio/wasmer/blob/master/examples/plugin.rs) we create a logging wrapper and redirect writes to stdout through it.

**Comment by cmyr** [comment]
@MarkMcCaskey sorry, missed this comment!

xi is on the back burner right now, but I am potentially interested in a plugin mechanism for [runebender](https://github.com/linebender/runebender); I'll definitely keep this in mind, I'm very curious.

**Comment by Cogitri** [comment]
I explored this for Tau too since I wanted to add hooks to it (e.g. run save.sh script during save, e.g. to run `cargo fmt`  when saving a rust file). It'd be nice that Gluon is platform agnostic, but I'd personally prefer just using SH due to it using a simpler syntax and more people knowing it.

**Comment by ShalokShalom** [comment]
SH has a simpler syntax as Gluon?

**Comment by Cogitri** [comment]
Well, it's certainly more well known

-------- Original Message --------

**Comment by ShalokShalom** [comment]
emacs is more well known as Xi

So by that standard..
